### PR TITLE
[sui framework] `address` and `hex` modules

### PR DESCRIPTION
**Pull request by sblackshear**
- Promote `hex` module used by capys to the framework
 - Add `address` module with useful functions for conversion:
     - Conversion of addresses to/from `u256` (useful if you want to compare them or do arithmetic operations)
     - Conversion of address to/from bytes (moved from `object` module)
     - Conversion of address to string

### CODE REVIEW

**Comment by sblackshear** [comment]
I am getting a stack overflow in`sui_cost::test_good_snapshot`--any ideas what might be causing this @oxade?

File: crates/sui-framework/sources/address.move [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+    public fun to_string(a: address): string::String {
+        string::from_ascii(to_ascii_string(a))
+    }
+
+    /// Length of a Sui address in bytes
+    public fun length(): u64 {
+        LENGTH
+    }
+
+    /// Largest possible address z
```
Comment (amnn) [inline crates/sui-framework/sources/address.move:54]: nit: Spurious `z`?

```suggestion
    /// Largest possible address
```

File: crates/sui-framework/sources/hex.move [hunk 2]
```diff
@@ -2,28 +2,34 @@
...
-        assert!(b"ff" == to_hex(x"ff"), 0);
-        assert!(b"fe" == to_hex(x"fe"), 0);
-        assert!(b"00" == to_hex(x"00"), 0);
+    fun test_hex_encode() {
+        assert!(b"30" == encode(b"0"), 0);
+        assert!(b"61" == encode(b"a"), 0);
+        assert!(b"ff" == encode(x"ff"), 0);
+        assert!(b"fe" == encode(x"fe"), 0);
+        assert!(b"00" == encode(x"00"), 0);
+        assert!(b"666666" == encode(b"fff"), 0);
```
Comment (amnn) [inline crates/sui-framework/sources/hex.move:33]: nit: Could you group together the tests where the string being encoded is a `b`yte string vs a he`x` string, and maybe add comments calling out the difference? The new `0`, `a` and `fff` made me double take without those (alternatively, have byte strings that couldn't also be interpreted as hex strings (but maybe that's the point of the test)?

File: crates/sui-framework/src/natives/address.rs [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+    loaded_data::runtime_types::Type, natives::function::NativeResult, pop_arg, values::Value,
+};
+use smallvec::smallvec;
+use std::collections::VecDeque;
+
+const E_ADDRESS_PARSE_ERROR: u64 = 0;
+
+const E_CANT_CONVERT_ADDRESS_TO_U256: u64 = 1;
+
+// Implementation of the Move native function address_from_bytes(bytes: vector<u8>): address;
```
Comment (amnn) [inline crates/sui-framework/src/natives/address.rs:18]: nit: comment needs updating.

```suggestion
// Implementation of the Move native function address::from_bytes(bytes: vector<u8>): address;
```

File: crates/sui-framework/src/natives/address.rs [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+
+    let addr = pop_arg!(args, AccountAddress);
+    let mut u256_bytes = [0; 32];
+    let mut addr_bytes_le = addr.to_vec();
+    addr_bytes_le.reverse();
+    // convert the address into a u256 by padding out the lower 12 bytes with 0's
+    for (i, b) in addr_bytes_le.into_iter().enumerate() {
+        u256_bytes[i] = b
+    }
+    let u256_val = Value::u256(U256::from_le_bytes(&u256_bytes));
```
Comment (amnn) [inline crates/sui-framework/src/natives/address.rs:56]: Could be simplified a little bit, I think, using `Vec::resize`:

```suggestion
    let mut addr_bytes_le = addr.to_vec();
    addr_bytes_le.reverse();
    addr_bytes_le.resize(32, 0);
    let u256_val = Value::u256(U256::from_le_bytes(&addr_bytes_le));
```

**Review by amnn — APPROVED** [review]
LGTM!

File: crates/sui-framework/sources/address.move [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+    public fun max(): u256 {
+        MAX
+    }
+}
```
Comment (tnowacki) [inline crates/sui-framework/sources/address.move:58]: 👀

**Review by tnowacki — COMMENTED** [review]
This is great! I really wish we had a module in the std for each primitive type (like we do with vector and signer). Even if they were just placeholders for the time being 

Though I suppose address is a bit weird in that it is different per instantiation

File: crates/sui-framework/sources/hex.move (hunk 2 above)
Comment (tnowacki) [reply crates/sui-framework/sources/hex.move:33]: Even after reading @amnn's comment here, it took me like 10 minutes to see that half of the tests are `encode(x"...")` and the other half are `encode(b"...")` lol 
Maybe two separate tests would be helpful

**Review by damirka — APPROVED** [review]

**Comment by sblackshear** [comment]
> This is great! I really wish we had a module in the std for each primitive type (like we do with vector and signer). Even if they were just placeholders for the time being
> 
> Though I suppose address is a bit weird in that it is different per instantiation

I agree--I think we should add one for each with (at least) `to_string` and `from_string`. Happy to upstream for the non-`address` types if the community likes it.

**Pull request merged**